lib/TPP/IR/StructuredOpMatcher.cpp

```diff
@@ -0,0 +1,330 @@
+    return fun(numDpsInputs);
+  });
+  return *this;
+}
+
+structured_match::StructuredOpMatcher &
+structured_match::StructuredOpMatcher::inputs(
+    structured_match::BinaryPredicate binaryPredicate) {
+  predicates.push_back([=](linalg::LinalgOp linalgOp) -> bool {
+    if (isa<_OR>(&binaryPredicate)) {
+      return (binaryPredicate.predicateOnLhs(linalgOp.getNumDpsInputs()) ||
+              binaryPredicate.predicateOnRhs(linalgOp.getNumDpsInputs()));
+    }
+    return false;
+  });
+  return *this;
+}
+
+structured_match::StructuredOpMatcher &
+structured_match::StructuredOpMatcher::hasBufferSemantics() {
+  predicates.push_back([](linalg::LinalgOp linalgOp) -> bool {
+    return linalgOp.hasBufferSemantics();
+  });
+  return *this;
+}
+
+structured_match::StructuredOpMatcher &
+structured_match::StructuredOpMatcher::hasTensorSemantics() {
+  predicates.push_back([](linalg::LinalgOp linalgOp) -> bool {
+    return linalgOp.hasTensorSemantics();
+  });
+  return *this;
+}
+
+structured_match::StructuredOpMatcher &
+structured_match::StructuredOpMatcher::verifyInterface(
+    std::function<LogicalResult(Operation *op)> fun) {
+  predicates.push_back([&](linalg::LinalgOp linalgOp) -> bool {
+    if (failed(fun(linalgOp)))
+      return false;
+    return true;
+  });
+  return *this;
+}
+
+//===---------------------------------------------------------------------===//
+// Input predicates.
+//===---------------------------------------------------------------------===//
+
+structured_match::StructuredOpMatcher &
+structured_match::StructuredOpMatcher::input(AllOperands tag, HasStaticShape) {
+  predicates.push_back([=](linalg::LinalgOp linalgOp) -> bool {
+    for (OpOperand *operand : linalgOp.getDpsInputOperands()) {
+      auto operandType = operand->get().getType();
+      if (auto shapedType = operandType.dyn_cast_or_null<ShapedType>())
+        if (!shapedType.hasStaticShape())
+          return false;
+    }
+    return true;
+  });
+  return *this;
+}
+
+structured_match::StructuredOpMatcher &
+structured_match::StructuredOpMatcher::input(
+    Operand operand, std::function<bool(AffineMap map)> fun) {
+  predicates.push_back([=](linalg::LinalgOp linalgOp) -> bool {
+    size_t idxOperand = operand.idx;
+    assert(idxOperand < static_cast<size_t>(linalgOp.getNumDpsInputs()));
+    AffineMap tiedIndexingMap = linalgOp.getMatchingIndexingMap(
+        linalgOp.getDpsInputOperand(idxOperand));
+    return fun(tiedIndexingMap);
+  });
+  return *this;
+}
+
+structured_match::StructuredOpMatcher &
+structured_match::StructuredOpMatcher::input(
+    AllOperands tag, std::function<bool(AffineMap map)> fun) {
+  predicates.push_back([=](linalg::LinalgOp linalgOp) -> bool {
+    for (OpOperand *operand : linalgOp.getDpsInputOperands()) {
+      AffineMap tiedIndexingMap = linalgOp.getMatchingIndexingMap(operand);
+      if (!fun(tiedIndexingMap))
+        return false;
+    }
+    return true;
+  });
+  return *this;
+}
```

Comment:
To expand on this a bit. I think the specific logic of accessing be it maps or shapes should be encoded in the functors you pass. `input` should focus on properly interpreting its operand tags but I feel like the rest of the current logic overspecializes the `input` functions.

Reply:
You would pass the functor the linalg operation? the way to get the affine map in linalg is pretty specific

Reply:
It could be an option if it allows us to easily create and pass new predicates.
But maybe affine maps are so specific that they require separate approach.

Current design just looks a bit inflexible.

Reply:
The problem I see here is that if you allows generic functors (i.e., that takes as input an op) the predicate is logic is more dictated by the functors rather than the matchers. For example, if the matcher wants to check an operand property the functors should ideally take only an opOperand as input, rather than the entire op.